### DISCUSSION TITLE
Best Practice: Selectively pulling specific layer types

### DISCUSSION THREAD

**Discussion by tcinbis** [oras-project/oras#444] [Q&A] [answered]
Hey everyone!

With oras-cli version pre `0.13.0` we were (maybe ab-) using the `--media-type` flag to instruct oras to only pull specific parts of an artifact. For example, we would publish a service artifact containing a container image, as well as, some configuration files and deployment scripts. Since in certain scenarios we would only need access to the scripts or configuration we could save time by just specifying a specific media type.

Because `--media-type` was removed with version `0.13.0` I am now wondering what the best solution would be to achieve a similar functionality? I saw that the OCI describes [Layer Types](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#layer-content-format). Could this help here? 

Thanks, everyone!

**Comment by qweeah** [comment]
Hi Tom, thanks for the question. Can you provide a more specific scenario on: you have a service artifact containing a container image and some configuration files but you only want to pull the configuration layers without the image layers.

For now, `oras pull` will only download layers that has names stored in [OCI title annotation](https://github.com/opencontainers/image-spec/blob/main/specs-go/v1/annotations.go#L52). So if the image layers are not annotated with title, then `oras pull` will skip them and only download the configuration layers.

From personal perspective, the best practice to pack a service artifact is not to put file layers and image layers all under one OCI artifact manifest, but to put them into different artifacts. We are preparing [0.14.0](https://github.com/oras-project/oras/milestone/6) and it supports a [new manifest type](https://github.com/oras-project/artifacts-spec#overview) to allow user attaching service files(e.g. service configurations) to a service image. After we released 0.14.0, you can download the configuration file layers via `oras pull <oras-artifact-reference>`

```
+---------------+            +---------------------+
| ORAS Artifact +------------> OCI Artifact (image)|
+-----+---------+            +-+---------+---------+
      |                        |         |
+-----v---------+        +-----v--+ +----v------------+
| File Layers...|+       | config | | Image Layers... |+
+---------------+|+      +--------+ +-----------------+|+
 +---------------+|                  +-----------------+|
  +---------------+                    +----------------+

```

For now, there is no plan to support more fine-grained layer filtering for `oras pull`

**Reply by qweeah** [reply]
Just synced with @Wwwsylvia and found that a new copy option, `FindSuccessors` is introduced to oras-go in [this PR](https://github.com/oras-project/oras-go/commit/9fc00616c490489943ad6a88b774d67e8e8c5a91). So there are some plans to support more fine-grained layer filtering in the future.

**Reply by caozhuozi** [reply]
Hi, @qweeah. How can we perform a fine-grained layer filtering now?

**Reply by Wwwsylvia** [reply]
Hi @caozhuozi , do you use ORAS CLI or SDK? Currently the Go SDK provides such functionality, but CLI doesn't.

**Reply by caozhuozi** [reply]
Hi, @Wwwsylvia  Thanks for looking at my question! 🙏
Could you please specify which part (or specific file) of the Go SDK relates to this option?

Thank you!

**Reply by Wwwsylvia** [reply, marked as answer]
The [`CopyGraphOptions.FindSuccessors`](https://pkg.go.dev/oras.land/oras-go/v2#CopyGraphOptions) callback function of `oras-go` allows you to customize the descriptors of layers to copy. Here is an example of how ORAS CLI uses this function:

https://github.com/oras-project/oras/blob/e89bc7b20df8f06d5386e46c090e9a08b0d88ebd/cmd/oras/root/attach.go#L191-L204

Feel free to let us know if you encounter any issues. You can also find us in the [ORAS slack channel](https://cloud-native.slack.com/archives/CJ1KHJM5Z).
